File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel.Tests/Cdm/Resolution/CircularResolutionTest.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using Microsoft.CommonDataModel.ObjectModel.Cdm;
using Microsoft.CommonDataModel.ObjectModel.Utilities;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;

namespace Microsoft.CommonDataModel.ObjectModel.Tests.Cdm.Resolution
{
    [TestClass]
    public class CircularResolutionTest
    {
        /// <summary>
        /// The path between TestDataPath and TestName.
        /// </summary>
        private string testsSubpath = Path.Combine("Cdm", "Resolution", "CircularResolution");

        /// <summary>
        /// Test proper behavior for entities that contain circular references
        /// </summary>
        [TestMethod]
        public async Task TestCircularReference()
        {
            CdmCorpusDefinition cdmCorpus = TestHelper.GetLocalCorpus(testsSubpath, "TestCircularReference");
            CdmEntityDefinition customer = await cdmCorpus.FetchObjectAsync<CdmEntityDefinition>("local:/Customer.cdm.json/Customer");
            CdmEntityDefinition resCustomerStructured = await customer.CreateResolvedEntityAsync("resCustomer", new ResolveOptions(
                customer.InDocument,
                new AttributeResolutionDirectiveSet(new HashSet<string> { "normalized", "structured", "noMaxDepth" })
            ));

            // check that the circular reference attribute has a single id attribute
            CdmAttributeGroupDefinition storeGroupAtt =
                ((CdmAttributeGroupReference)resCustomerStructured.Attributes[1]).ExplicitReference as CdmAttributeGroupDefinition;
            CdmAttributeGroupDefinition customerGroupAtt =
                ((CdmAttributeGroupReference)storeGroupAtt.Members[1]).ExplicitReference as CdmAttributeGroupDefinition;
            Assert.AreEqual(1, customerGroupAtt.Members.Count);
            Assert.AreEqual("customerId", ((CdmTypeAttributeDefinition)customerGroupAtt.Members[0]).Name);
        }

        /// <summary>
        /// Test that relationship is created when an entity contains a reference to itself
        /// </summary>
        [TestMethod]
        public async Task TestSelfReference()
        {
            CdmCorpusDefinition cdmCorpus = TestHelper.GetLocalCorpus(testsSubpath, "TestSelfReference");
            CdmManifestDefinition manifest = await cdmCorpus.FetchObjectAsync<CdmManifestDefinition>("local:/SelfReference.manifest.cdm.json");
            await cdmCorpus.CalculateEntityGraphAsync(manifest);
            await manifest.PopulateManifestRelationshipsAsync();

            Assert.AreEqual(1, manifest.Relationships.Count);
            CdmE2ERelationship rel = manifest.Relationships[0];
            Assert.AreEqual("CustTable.cdm.json/CustTable", rel.FromEntity);
            Assert.AreEqual("CustTable.cdm.json/CustTable", rel.ToEntity);
            Assert.AreEqual("FactoringAccountRelationship", rel.FromEntityAttribute);
            Assert.AreEqual("PaymTermId", rel.ToEntityAttribute);
        }
    }
}

File: objectModel/TypeScript/__test__/Cdm/Resolution/CircularResolution.test.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { CdmAttributeGroupDefinition } from '../../../Cdm/CdmAttributeGroupDefinition';
import { CdmAttributeGroupReference } from '../../../Cdm/CdmAttributeGroupReference';
import { CdmCorpusDefinition } from '../../../Cdm/CdmCorpusDefinition';
import { CdmE2ERelationship } from '../../../Cdm/CdmE2ERelationship';
import { CdmEntityDefinition } from '../../../Cdm/CdmEntityDefinition';
import { CdmManifestDefinition } from '../../../Cdm/CdmManifestDefinition';
import { CdmTypeAttributeDefinition } from '../../../Cdm/CdmTypeAttributeDefinition';
import { AttributeResolutionDirectiveSet } from '../../../Utilities/AttributeResolutionDirectiveSet';
import { resolveOptions } from '../../../Utilities/resolveOptions';
import { testHelper } from '../../testHelper';

describe('Cdm/Resolution/CircularResolution', () => {
    /**
     * The path between TestDataPath and TestName.
     */
    const testsSubpath: string = 'Cdm/Resolution/CircularResolutionTest';

    /**
     * Test proper behavior for entities that contain circular references
     */
    it('TestCircularReference', async () => {
        const cdmCorpus: CdmCorpusDefinition = testHelper.getLocalCorpus(testsSubpath, 'TestCircularReference');
        const customer: CdmEntityDefinition = await cdmCorpus.fetchObjectAsync<CdmEntityDefinition>('local:/Customer.cdm.json/Customer');
        const resCustomerStructured: CdmEntityDefinition =
            await customer.createResolvedEntityAsync('resCustomer', new resolveOptions(
                customer.inDocument,
                new AttributeResolutionDirectiveSet(new Set<string>(['normalized', 'structured', 'noMaxDepth']))
            ));

        // check that the circular reference attribute has a single id attribute
        const storeGroupAtt: CdmAttributeGroupDefinition =
            (resCustomerStructured.attributes.allItems[1] as CdmAttributeGroupReference).explicitReference as CdmAttributeGroupDefinition;
        const customerGroupAtt: CdmAttributeGroupDefinition =
            (storeGroupAtt.members.allItems[1] as CdmAttributeGroupReference).explicitReference as CdmAttributeGroupDefinition;
        expect(customerGroupAtt.members.length)
            .toBe(1);
        expect((customerGroupAtt.members.allItems[0] as CdmTypeAttributeDefinition).name)
            .toBe('customerId');
    });

    /**
     * Test that relationship is created when an entity contains a reference to itself
     */
    it('TestSelfReference', async () => {
        const cdmCorpus: CdmCorpusDefinition = testHelper.getLocalCorpus(testsSubpath, 'TestSelfReference');
        const manifest: CdmManifestDefinition = await cdmCorpus.fetchObjectAsync<CdmManifestDefinition>('local:/SelfReference.manifest.cdm.json');
        await cdmCorpus.calculateEntityGraphAsync(manifest);
        await manifest.populateManifestRelationshipsAsync();

        expect(manifest.relationships.length)
            .toBe(1);
        const rel: CdmE2ERelationship = manifest.relationships.allItems[0];
        expect(rel.fromEntity)
            .toBe('CustTable.cdm.json/CustTable');
        expect(rel.toEntity)
            .toBe('CustTable.cdm.json/CustTable');
        expect(rel.fromEntityAttribute)
            .toBe('FactoringAccountRelationship');
        expect(rel.toEntityAttribute)
            .toBe('PaymTermId');
    });
});
